apps/alerts/main.py

```diff
@@ -34,7 +54,52 @@ def index(request):
         if request.path == "/":
             return main_html
 
-    except:
+        if request.path.endswith("list_exams"):
+            return jsonify(db.collection("exam-alerts").document("all").get().to_dict()["exam-list"])
+
+        if request.path.endswith("fetch_data"):
+            exam = request.json["exam"]
+            email = get_email(request)
+            student_data = db.collection("exam-alerts").document(exam).collection("students").document(email).get().to_dict()
+            announcements = list(db.collection("exam-alerts").document(exam).collection("announcements").stream())
+            return jsonify({
+                "success": True,
+                "exam_type": "ok-exam",
+                "questions": [
+                    {
+                        "questionName": question["student_question_name"],
+                        "startTime": question["start_time"],
+                        "endTime": question["end_time"],
+                    }
+                    for question in student_data["questions"]
+                ],
+                "announcements": get_announcements(student_data, announcements),
+            })
+
+        if request.path.endswith("clear_announcements"):
+            exam = request.json["exam"]
+            course = exam.split("-")[0]
+            if not is_admin(get_email(request), course):
+                abort(401)
+            clear_collection(db, db.collection("exam-alerts").document(exam).collection("announcements"))
+            return {"success": True}
+
+        if request.path.endswith("add_announcement"):
+            exam = request.json["exam"]
+            course = exam.split("-")[0]
+            if not is_admin(get_email(request), course):
+                abort(401)
+            announcement = request.json["announcement"]
+            announcement["timestamp"] = time.time()
+            db.collection("exam-alerts").document(exam).collection("announcements").document().set(announcement)
```

Comment:
this appends I assume?

Reply:
it creates a new document in the collection, yes. Collections are unordered AFAIK